druid/src/widget/view_switcher.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A widget that can dynamically switch between one of many views.
+
+use crate::{
+    BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx,
+    Point, Rect, Size, UpdateCtx, Widget, WidgetPod,
+};
+
+/// A widget that can switch dynamically between one of many views depending
+/// application state.
+pub struct ViewSwitcher<T: Data, U: PartialEq> {
+    child_selector: Box<dyn Fn(&T, &Env) -> U>,
+    child_builder: Box<dyn Fn(&U, &Env) -> Box<dyn Widget<T>>>,
+    active_child: Option<WidgetPod<T, Box<dyn Widget<T>>>>,
```

Comment:
I don't know if it's worth forcing a user to specify a default widget explicitly, since that case can be handled inside `child_builder` anyway. I'm working with the assumption that in most applications, the value returned by `child_picker` will always -- or almost always -- correspond to a valid widget inside `child_builder`.

If we explicitly force the user to specify a default widget, and in their application there's never really a case when `child_picker` returns something that `child_builder` cannot understand, then allocating that extra default widget might be a waste. Plus, it's extra cognitive overhead to think of a reasonable default widget when you're absolutely certain that every value returned by `child_picker` can be understood by `child_builder`.

I hope that made sense.

Reply:
sounds reasonable. :)